File: EngineKitGame/EngineKit/EngineKit/Items/Shape/Cone.m


#import "Cone.h"


@implementation Cone

+ (instancetype)cone {
    return [self new];
}

- (instancetype)init {
    if (self = [super initAndAddToScene]) {
        self.cone = [SCNCone new];
    }
    return self;
}

- (instancetype)initWithTopRadius:(CGFloat)topRadius
                     bottomRadius:(CGFloat)bottomRadius
                           height:(CGFloat)height {
    if (self = [super initAndAddToScene]) {
        self.cone = [SCNCone coneWithTopRadius:topRadius
                                  bottomRadius:bottomRadius
                                        height:height];
    }
    return self;
}

- (Item *)deepCopy {
    Shape *newItem = [[self class] new];
    [self copyInfoTo:newItem];
    [super copyPhysicsTo:newItem];
    return newItem;
}

- (void)copyInfoTo:(Cone *)item {
    [super copyInfoTo:item];
    
    item.bottomRadius = self.bottomRadius;
    item.topRadius = self.topRadius;
    item.height = self.height;
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark - Property Overriding

- (void)setCone:(SCNCone *)cone {
    self.geometry = cone;
}

- (SCNCone *)cone {
    return (SCNCone *)self.geometry;
}


- (void)setRadius:(NSNumber *)radius {
    [self assertTheresNoPhysicsBody];
    self.cone.bottomRadius = radius.doubleValue;
}

- (NSNumber *)radius {
    return @(self.cone.bottomRadius);
}

- (void)setBottomRadius:(NSNumber *)radius {
    [self assertTheresNoPhysicsBody];
    self.radius = radius;
}

- (NSNumber *)bottomRadius {
    return self.radius;
}

- (void)setTopRadius:(NSNumber *)radius {
    [self assertTheresNoPhysicsBody];
    self.cone.topRadius = radius.doubleValue;
}

- (NSNumber *)topRadius {
    return @(self.cone.topRadius);
}

- (void)setHeight:(NSNumber *)height {
    [self assertTheresNoPhysicsBody];
    self.cone.height = height.doubleValue;
}

- (NSNumber *)height {
    return @(self.cone.height);
}

@end

File: Project/Project/Items/Shapes/Cone.swift

import Foundation
import SceneKit
import JavaScriptCore



extension SCNCone {
    var size: CGFloat {
        set {
            
            if bottomRadius != 0 {
                let ratio = topRadius / bottomRadius
                
                if topRadius > bottomRadius {
                    topRadius = newValue / 2
                    bottomRadius = topRadius / ratio
                }
                else {
                    bottomRadius = newValue / 2
                    topRadius = bottomRadius * ratio
                }
            }
            else if topRadius != 0 {
                let ratio = bottomRadius / topRadius
                
                if topRadius > bottomRadius {
                    topRadius = newValue / 2
                    bottomRadius = topRadius * ratio
                }
                else {
                    bottomRadius = newValue / 2
                    topRadius = bottomRadius / ratio
                }
            }
            
            height = newValue
        }
        get {
            return max(max(bottomRadius * 2, topRadius * 2), height)
        }
    }
    
    var radius: CGFloat {
        set { bottomRadius = newValue }
        get { return bottomRadius     }
    }
}



@objc class Cone: Shape {
    
    
    var radius: CGFloat {
        set { cone.radius = newValue }
        get { return cone.radius     }
    };  var bottomRadius: CGFloat {
        set { cone.bottomRadius = newValue }
        get { return cone.bottomRadius     }
    }
    
    var topRadius: CGFloat {
        set { cone.topRadius = newValue }
        get { return cone.topRadius     }
    }
    
    var height: CGFloat {
        set { cone.height = newValue }
        get { return cone.height     }
    }
    
    var cone: SCNCone {
        set { geometry = newValue        }
        get { return geometry as SCNCone }
    }
    
    override init() {
        super.init(geometry: SCNCone())
    }
}